experimental/dds/tree2/src/util/utils.ts

```diff
@@ -218,3 +219,17 @@ export function isJsonObject(
 ): value is { readonly [P in string]?: JsonCompatibleReadOnly } {
 	return typeof value === "object" && value !== null && !Array.isArray(value);
 }
+
+export function assertValidIndex(
+	index: number,
+	array: { length: number },
```

Comment:
I think this is why you had to make a bunch of changes that should have been unnecessary
```suggestion
	array: { readonly length: number },
```

Reply:
changed to readonly